walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,60 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
 
 public final class FindMeetingQuery {
+  /* Finds all time ranges that satisfy the request given the already existing set of events. */
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    // Find potential conflicts between scheduled events and request.
+    List<TimeRange> potentialConflicts = getPotentialConflicts(events, request);
+
+    // Sort the potential conflicts in order of ascending start time.
+    Collections.sort(potentialConflicts, TimeRange.ORDER_BY_START);
+   
+    Collection<TimeRange> timeRanges = new ArrayList<TimeRange>();
+    int lastConflictEnd = TimeRange.START_OF_DAY;
+    long requestDuration = request.getDuration();
+    for (TimeRange currentConflict : potentialConflicts) {
+        // If the time between the end time of the current conflict and the latest end time 
+        // conflict seen so far is greater that the request duration provided,
+        // add a new time range to the time ranges returned.
+        int currentConflictStart = currentConflict.start();
+        if (currentConflictStart - lastConflictEnd >= requestDuration) {
+            timeRanges.add(TimeRange.fromStartEnd(lastConflictEnd, currentConflictStart, /* inclusive= */ false));
+        }
+
+        // Update the latest conflict end time seen so far.
+        int currentConflictEnd = currentConflict.end();
+        if (currentConflictEnd > lastConflictEnd)
+            lastConflictEnd = currentConflictEnd;
+    }
+
+    // Check if the time between the latest end time and the end of the day is a valid time range.
+    if (TimeRange.END_OF_DAY - lastConflictEnd >= requestDuration) {
+        timeRanges.add(TimeRange.fromStartEnd(lastConflictEnd, TimeRange.END_OF_DAY, /* inclusive= */ true));
+    }
+    return timeRanges;
+  }
+
+  private static List<TimeRange> getPotentialConflicts(Collection<Event> events, MeetingRequest request) {
+    List<TimeRange> potentialConflicts = new ArrayList<TimeRange>();
+    Collection<String> requestAttendees = request.getAttendees();
+    for (Event event : events) {
+        // Find the intersection between current event attendees and request attendees.
+        HashSet<String> intersection = new HashSet<String>(event.getAttendees());
```

Comment:
I think usually you'd declare a HashSet (or LinkedHashSet or whatever else implements Set) as a Set to keep it as generic as possible

Reply:
Done. Noted.